### PR TITLE
[RISCV] Add a check in lowerSELECT after foldBinOpIntoSelectIfProfitable

### LINKED ISSUE #97390: [RISCV] Node returned by foldBinOpIntoSelectIfProfitable reports error in lowerSELECT
Here is a simple reduced case:
`int a = (a ? 4 : 8) / (8 ? -5141143369814759789L : 0);`
When compiling this case with order 
`clang++  -std=c++11 -march=rv64imafdc -mabi=lp64d -w -O0 -o a.o -c a.cpp`
It reports error: llvm-project/llvm/include/llvm/CodeGen/SelectionDAGNodes.h:960: const llvm::SDValue& llvm::SDNode::getOperand(unsigned int) const: Assertion `Num < NumOperands && "Invalid child # of SDNode!"' failed.

It is because for certain cases foldBinOpIntoSelectIfProfitable may return a constant node when trueVal and falseVal of the select inst have the same value, at this time in following lowerSELECT it reports error that operands cannot be get from a constant node.

### PR DESCRIPTION
**Pull request by Yunzezhu94**
In certain case foldBinOpIntoSelectIfProfitable may return a constant node, the node will be lowered in lowerSELECT and lead to crash.
This patch fix the bug by adding an extra check before lowerSELECT that do lowerSELECT as before when foldBinOpIntoSelectIfProfitable returns a select node, and return the node directly when foldBinOpIntoSelectIfProfitable returns a constant node.

Fixes https://github.com/llvm/llvm-project/issues/97390

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/fold-binop-into-select.ll [hunk 1]
```diff
@@ -58,3 +58,15 @@ entry:
   %res = sub i64 2, %select_
   ret i64 %res
 }
+
+define i64 @fold_binop_into_select_return_constant(i1 %c) {
```
Comment (dtcxzyw) [inline llvm/test/CodeGen/RISCV/fold-binop-into-select.ll:62]: Please make sure that this function cannot be folded by middle-end.
See: https://godbolt.org/z/h1M7frG4j

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/test/CodeGen/RISCV/fold-binop-into-select.ll (hunk 1 above)
Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/fold-binop-into-select.ll:62]: Sorry, it is fine for crash fixes.

File: llvm/test/CodeGen/RISCV/fold-binop-into-select.ll [hunk 2]
```diff
@@ -58,3 +58,15 @@ entry:
...
+; CHECK-LABEL: fold_binop_into_select_return_constant:
+; CHECK:       # %bb.0: # %entry
+; CHECK-NEXT:    li a0, 0
+; CHECK-NEXT:    ret
+entry:
```
Comment (dtcxzyw) [inline llvm/test/CodeGen/RISCV/fold-binop-into-select.ll:67]: Sorry, I cannot reproduce this crash with a8e1c3e1239604ac787b6a2d39b5278ddec8aa8a.

File: llvm/test/CodeGen/RISCV/fold-binop-into-select-return-constant.ll [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+; CHECK-LABEL: fold_binop_into_select_return_constant:
+; CHECK:       # %bb.0: # %entry
+; CHECK-NEXT:    li a0, 0
+; CHECK-NEXT:    ret
+entry:
+  %select1 = select i1 %c, i32 4, i32 8
+  %select2 = sext i32 %select1 to i64
+  %div1 = sdiv i64 %select2, -5141143369814759789
+  ret i64 %div1
+}
```
Comment (dtcxzyw) [inline llvm/test/CodeGen/RISCV/fold-binop-into-select-return-constant.ll:13]: Missing newline.

File: llvm/test/CodeGen/RISCV/fold-binop-into-select-return-constant.ll [hunk 4]
```diff
@@ -0,0 +1,13 @@
+; RUN: llc -mtriple=riscv64 -mattr=+m < %s  | FileCheck  %s
```
Comment (dtcxzyw) [inline llvm/test/CodeGen/RISCV/fold-binop-into-select-return-constant.ll:1]: Please regenerate check lines with `llvm/utils/update_llc_test_checks.py`.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Nice catch!

File: llvm/test/CodeGen/RISCV/fold-binop-into-select-return-constant.ll (hunk 4 above)
Comment (Yunzezhu94) [reply llvm/test/CodeGen/RISCV/fold-binop-into-select-return-constant.ll:2]: Updated.

**Review by dtcxzyw — APPROVED** [review]
Thank you!

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 5]
```diff
@@ -7664,7 +7664,10 @@ SDValue RISCVTargetLowering::lowerSELECT(SDValue Op, SelectionDAG &DAG) const {
       if (SDValue NewSel = foldBinOpIntoSelectIfProfitable(*Op->use_begin(),
                                                            DAG, Subtarget)) {
         DAG.ReplaceAllUsesWith(BinOp, &NewSel);
-        return lowerSELECT(NewSel, DAG);
+        if (NewSel.getOpcode() == ISD::SELECT)
+          return lowerSELECT(NewSel, DAG);
+        else
+          return NewSel;
```
Comment (mgudim) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:7670]: I think using ternary operator is preferred in LLVM

**Comment by mgudim** [comment]
@Yunzezhu94 thanks for working on this, looks good. I left some minor comments.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 5 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:7670]: The preferred thing would be to do

```
if (NewSel.getOpcode() == ISD::SELECT)
  return lowerSELECT(NewSel, DAG);
return NewSel;
```

**Comment by topperc** [comment]
The title and the description don't say what the extra check is.

**Comment by mgudim** [comment]
> The title and the description don't say what the extra check is.

The description says what the check is.

**Comment by topperc** [comment]
> > The title and the description don't say what the extra check is.
> 
> The description says what the check is.

It could be a little better in my opinion. The first sentence says that `foldBinOpIntoSelectIfProfitable` returned a constant node. The second sentence says a new check is added. Reading the two sentences together, you might think that the new check is that the node isn't a constant. But that's not exactly what the new check is.

**Comment by Yunzezhu94** [comment]
Updated code and description as advised.

**Comment by topperc** [comment]
> Updated code and description as advised.

I think you only updated the commit message. You need to update the PR description too. The "squash and merge" will use the PR description not the commit message.

**Review by mgudim — APPROVED** [review]
LGTM, thanks!

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**